### PR TITLE
Group remove members

### LINKED ISSUE #335: Allow removing users from a group
## Motivations
Allow removing users from a group

## Subtasks
In the group composition tab, direct or all descendants, for "users":
- add a check box in front of the row to allow the selection of rows
- add a "trash" icon + "Remove from group" below-right of the table, which is greyed out when there is no selection
- on click, call [groupRemoveMembers](https://france-ioi.github.io/algorea-devdoc/api/#operation/groupRemoveMembers) with the list of user ids to remove. While loading, freeze the table and actions (grey out + no click on the remove button).
- on success/failure/partial success, do the same done in pending request component

Double-Triple check all behaviors work as expected before asking for the review.

### PR DESCRIPTION
**Pull request by GabrielVidal1**
Fixes #335 

As previously said during a meeting, a lot of the code wrapping calls of the MessageService is juste the same as the code in [pending-request-response-handling.ts](https://github.com/France-ioi/AlgoreaFrontend/blob/a24e1114a40b50f87cfec325fa57afd2021d0890/src/app/modules/group/components/pending-request/pending-request-response-handling.ts). All this code can be centralized in a different PR.

### CODE REVIEW

**Review by smadbe — CHANGES_REQUESTED** [review]
Bug: 
- After removing a user, the "remove button" stay selected, if I repress it I get a failure
- The same 
  - I select one user, remove it. Ok
  - I select another user, remove it. I get a response that one has been deleted and one has failed (while there only one selected)
So removed users stay selected. 
Please (manual) test your implementation in details before asking for review !!!

**Review by smadbe — CHANGES_REQUESTED** [review]
To be fixed: The checkbox+remove button are closer to the next component than the table. 
Screenshot 2021-05-04 at 15 08 07

File: src/app/modules/group/components/member-list/member-list.component.ts [hunk 1]
```diff
@@ -86,8 +86,7 @@ export class MemberListComponent implements OnChanges, OnDestroy {
     private messageService: MessageService,
   ) {
     this.dataFetching.pipe(
-      switchMap(params => this.getData(params.groupId, params.filter, params.sort)),
-      mapToFetchState(),
+      switchMap(params => this.getData(params.groupId, params.filter, params.sort).pipe(mapToFetchState())),
```
Comment (smadbe) [inline src/app/modules/group/components/member-list/member-list.component.ts:89]: Why?

**Review by smadbe — CHANGES_REQUESTED** [review]

File: src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts [hunk 2]
```diff
@@ -1,11 +1,6 @@
 import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
 
-export enum TypeFilter {
-  Groups,
-  Sessions,
-  Teams,
-  Users,
-}
+export type TypeFilter = 'groups' | 'sessions' | 'teams' | 'users';
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts:3]: fixed

File: src/app/modules/group/http-services/group-users.service.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import { ActionResponse, objectToMap, successData } from 'src/app/shared/http-services/action-response';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class GroupUsersService {
+
+  constructor(private http: HttpClient) {}
+
+  removeUsers(groupId: string, usersId: string[]): Observable<Map<string, any>> {
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/http-services/group-users.service.ts:15]: fixed, I also added this to the `RequestActionsService`, which uses the same enum

File: src/app/modules/group/components/member-list/user-removal-response-handling.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import { ERROR_MESSAGE } from 'src/app/shared/constants/api';
+import { TOAST_LENGTH } from 'src/app/shared/constants/global';
+
+export interface Result {
+  countRequests: number;
+  countSuccess: number;
+}
+
+export function displayResponseToast(messageService: MessageService, result: Result): void {
+  if (result.countSuccess === result.countRequests) {
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/user-removal-response-handling.ts:11]: fixed

File: src/app/modules/group/components/member-list/member-list.component.html [hunk 5]
```diff
@@ -38,6 +42,9 @@
       let-columns="columns"
     >
       <tr>
+        <td *ngIf="currentFilter.type === 'users'" style="width: 4rem; text-overflow: clip;">
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.html]: fixed

File: src/app/modules/group/components/member-list/member-list.component.scss [hunk 6]
```diff
@@ -59,6 +59,48 @@
         text-align: center;
         font-size: medium;
       }
+
+    }
+
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.scss:9]: fixed, I reworked the scss for the summary to harmonize it across all the tables.

File: src/app/modules/group/components/member-list/member-list.component.ts [hunk 7]
```diff
@@ -184,9 +189,18 @@ export class MemberListComponent implements OnChanges, OnDestroy {
...
+        displayResponseToast(this.messageService, parseResults(result));
+        this.table?.clear();
+        if (this.group) {
+          this.dataFetching.next({ groupId: this.group.id, filter: this.currentFilter, sort: this.currentSort });
+        }
+      },
+      _err => {
+        this.state = 'ready';
+        processRequestError(this.messageService);
+      });
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.ts]: I added a fourth state 'processing' for ongoing removals, just like in the pending request component, I don't know if it's the best way to handle it (this or having an other observable just for the removal)

File: src/app/modules/group/components/member-list/member-list.component.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/modules/group/components/member-list/member-list.component.scss:9]: there are still 2 new `ng-deep` in the PR while ng-deep should not be used at all.

[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/modules/group/components/member-list/member-list.component.scss:9]: You can ask directly "Александр Семенов " on Slack what's the best way to do in this case

File: src/app/modules/group/components/member-list/member-list.component.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/modules/group/components/member-list/member-list.component.ts]: That's not much better. What I had in mind is to use the usual FetchState for fetching the list, and another state which can just be a boolean "removalInProgress". Both can be observable.

File: src/app/modules/group/components/member-list/member-list.component.ts (hunk 1 above)
Comment (smadbe) [reply src/app/modules/group/components/member-list/member-list.component.ts:90]: Have you tried the resetter at the end?

File: src/app/modules/group/components/member-list/member-list.component.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.ts]: fixed

File: src/app/modules/group/components/member-list/member-list.component.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.scss]: fixed

File: src/app/modules/group/components/member-list/member-list.component.ts (hunk 1 above)
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.ts:91]: I tried to pass the `dataFetching` Subjet as an observable but le table is stuck on loading, so I kept it like it is.

File: src/app/modules/group/components/member-list/member-list.component.html [hunk 8]
```diff
@@ -77,7 +77,7 @@
         </span>
         <span class="filler"></span>
         <div class="summary-actions">
-          <span (click)="onRemove()" [ngClass]="{disabled: state !== 'ready' || selection.length === 0}">
+          <span (click)="onRemove()" [ngClass]="{disabled: (removalInProgress$ | async) || state !== 'ready' || selection.length === 0}">
```
Comment (smadbe) [inline src/app/modules/group/components/member-list/member-list.component.html:80]: `[ngClass]="{disabled: ...}"` can be replaced by `[class.disabled] = ".."`, shorter/cleaner in this case.

File: src/app/modules/group/components/member-list/member-list.component.ts [hunk 9]
```diff
@@ -60,7 +60,7 @@ export class MemberListComponent implements OnChanges, OnDestroy {
 
   @Input() group? : Group;
 
-  state: 'error' | 'ready' | 'fetching' | 'processing' = 'fetching';
+  state: 'error' | 'ready' | 'fetching' = 'fetching';
```
Comment (smadbe) [inline src/app/modules/group/components/member-list/member-list.component.ts]: is there a reason not to use fetchstate?

**Review by smadbe — CHANGES_REQUESTED** [review]
We are converging 👍

File: src/app/modules/group/components/member-list/member-list.component.ts (hunk 9 above)
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.ts]: With the fetchstate version, the table is empty during loading, and the data isn't stored anymore in the component, but it's needed for selection.

File: src/app/modules/group/components/member-list/member-list.component.html (hunk 8 above)
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.html]: fixed

File: src/app/modules/group/components/member-list/member-list.component.html [hunk 10]
```diff
@@ -77,7 +77,7 @@
         </span>
         <span class="filler"></span>
         <div class="summary-actions">
-          <span (click)="onRemove()" [ngClass]="{disabled: (removalInProgress$ | async) || state !== 'ready' || selection.length === 0}">
+            <span (click)="onRemove()" [class.disabled]="(removalInProgress$ | async) || state === 'ready' || selection.length === 0">
```
Comment (smadbe) [inline src/app/modules/group/components/member-list/member-list.component.html:80]: `state !== 'ready'` -> `=== 'ready'` ?  🤨

**Review by smadbe — CHANGES_REQUESTED** [review]

File: src/app/modules/group/components/member-list/member-list.component.html (hunk 10 above)
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.html]: fixed

**Review by smadbe — APPROVED** [review]

**Pull request merged**